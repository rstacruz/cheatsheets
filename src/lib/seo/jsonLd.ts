import type { JsonLdDocument } from '~/types/JsonLdDocument'
import { site } from '../../config'
import { getDescription, getPageImage, getPageURL } from './seo'

export function getJSONLDsForPage(page: {
  // TODO: rename getJsonLdSchemasForPage
  slug: string
  frontmatter: {
    title?: string
    description?: string
    category?: string
  }
}): Array<JsonLdDocument> {
  const description = getDescription(page)
  const image = getPageImage(page)
  const url = getPageURL(page)
  const category = page.frontmatter.category ?? 'Others'
  const categoryAnchor = category.toLowerCase().replace(/ /g, '-')
  const headline = page.frontmatter.title
    ? `${page.frontmatter.title} cheatsheet`
    : ''

  const newsArticle: JsonLdDocument = {
    '@context': 'http://schema.org',
    '@type': 'NewsArticle',
    mainEntityOfPage: {
      '@type': 'WebPage',
      '@id': 'https://google.com/article'
    },
    headline,
    image: [image],
    description: description
  }

  const breadcrumb: JsonLdDocument = {
    '@context': 'http://schema.org',
    '@type': 'BreadcrumbList',
    itemListElement: [
      {
        '@type': 'ListItem',
        position: 1,
        item: {
          '@id': `${site.url}/#${categoryAnchor}`,
          name: category
        }
      },
      {
        '@type': 'ListItem',
        position: 2,
        item: {
          '@id': url,
          name: headline
        }
      }
    ]
  }

  return [newsArticle, breadcrumb]
}
