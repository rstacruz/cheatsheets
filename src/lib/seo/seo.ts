import snarkdown from 'snarkdown'
import { site } from '../../config'
import type { SheetPage } from '../page'

export function getSEOPropsForHome() {
  const t = {
    title: 'Devhints — TL;DR for developer documentation',
    description: 'A ridiculous collection of web development cheatsheets'
  }
  const url = site.url
  const image = 'https://assets.devhints.io/previews/index.jpg'

  return {
    title: t.title,
    links: {
      canonical: url
    },
    meta: {
      description: t.description,
      'app:pageurl': url
    },
    metaProperties: denull({
      'og:description': t.description,
      'og:image:height': '471',
      'og:image': image,
      'og:image:width': '900',
      'og:site_name': site.title,
      'og:title': t.title,
      'og:type': 'website',
      'og:url': url,
      // BUG: twitter card props should be metaNames
      'twitter:title': image,
      'twitter:image': image,
      'twitter:description': t.description
    })
  }
}

export function getSEOPropsForPage(
  page: Pick<SheetPage, 'slug' | 'frontmatter'>
) {
  const title = `${page.frontmatter.title} cheatsheet`
  const description = getDescription(page)
  const image = getPageImage(page)
  const url = getPageURL(page)

  return {
    title: title,
    links: {
      canonical: url
    },
    meta: {
      description,
      'app:pageurl': url
    },
    metaProperties: denull({
      'og:description': description,
      'og:image:height': '471',
      'og:image': image,
      'og:image:width': '900',
      'og:site_name': site.title,
      'og:title': title,
      'og:type': 'article',
      'og:url': url,
      'article:tag': page.frontmatter.tags ?? [],
      'article:section': page.frontmatter.category,
      // BUG: twitter card props should be metaNames
      'twitter:title': title,
      'twitter:image': image,
      'twitter:description': description
    })
  }
}

/**
 * Return a description for a page
 */

export function getDescription(page: {
  frontmatter: {
    title?: string
    intro?: string
    description?: string
    keywords?: string[]
  }
}) {
  const t = {
    withDescriptionAndIntro: `{description} {intro}`,
    withDescription: `{description} · One-page guide to {title}`,
    withKeywordsAndIntro: `{keywords} · {intro}`,
    withKeywords: `{keywords} · One-page guide to {title}`,
    withIntro: `One-page guide to {title}: usage, examples, and more. {intro}`,
    default: `The one-page guide to {title}: usage, examples, links, snippets, and more.`
  }

  let fmt: string = t.default

  if (page.frontmatter.description && page.frontmatter.intro) {
    fmt = t.withDescriptionAndIntro
  } else if (page.frontmatter.description) {
    fmt = t.withDescription
  } else if (page.frontmatter.keywords && page.frontmatter.intro) {
    fmt = t.withKeywordsAndIntro
  } else if (page.frontmatter.keywords) {
    fmt = t.withKeywords
  } else if (page.frontmatter.intro) {
    fmt = t.withIntro
  } else {
    fmt = t.default
  }

  return fmt
    .replace('{title}', () => page.frontmatter.title ?? '')
    .replace('{keywords}', () => (page.frontmatter.keywords ?? []).join(' · '))
    .replace('{intro}', () => toPlainText(page.frontmatter.intro ?? ''))
    .replace('{description}', () =>
      toPlainText(page.frontmatter.description ?? '')
    )
}

/**
 * Remove falsy values from an object
 */

function denull(
  record: Record<string, string | string[] | null | undefined>
): Record<string, string | string[]> {
  return Object.fromEntries(
    Object.entries(record).filter((entry): entry is [string, string] =>
      Boolean(entry[1])
    )
  )
}

export function getPageImage({ slug }: { slug: string }) {
  return `https://assets.devhints.io/previews/${slug}.jpg`
}

export function getPageURL({ slug }: { slug: string }) {
  return new URL(`/${slug}`, site.url).toString()
}

/**
 * Convert Markdown to plain text.
 */

export function toPlainText(input: string) {
  const html = snarkdown(input)
  const plainText = html.replace(/<[^>]*>/g, '').replace(/\n/g, ' ')
  return plainText
}
