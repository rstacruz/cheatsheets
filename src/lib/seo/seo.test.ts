import { getSEOPropsForPage, toPlainText } from './seo'

const slug = 'react'
const title = 'React'
const description = 'A React cheatsheet'
const markdownDescription =
  'A [React](https://react.dev) cheatsheet (*Markdown text*)'
const keywords = ['hooks', 'components', 'props']

describe('getSEOPropsForPage()', () => {
  runTest({
    title: 'title only',
    input: { slug, frontmatter: { title } }
  })
  runTest({
    title: 'with description',
    input: { slug, frontmatter: { title, description } }
  })
  runTest({
    title: 'with Markdown description',
    input: { slug, frontmatter: { title, description: markdownDescription } }
  })
  runTest({
    title: 'with description, keywords',
    input: { slug, frontmatter: { title, description, keywords } }
  })
  runTest({
    title: 'with keywords',
    input: { slug, frontmatter: { title, keywords } }
  })

  function runTest({ title, input }: { title: string; input: any }) {
    test(title, () => {
      const output = getSEOPropsForPage(input)
      expect(output).toMatchSnapshot()
    })
  }
})

describe('toPlainText()', () => {
  test('converts markdown to plain text', () => {
    const input = 'This is a **bold** text'
    const output = toPlainText(input)
    expect(output).toEqual('This is a bold text')
  })

  test('removes HTML tags', () => {
    const input = '<p>This is a <strong>bold</strong> text</p>'
    const output = toPlainText(input)
    expect(output).toEqual('This is a bold text')
  })

  test('handles empty input', () => {
    const input = ''
    const output = toPlainText(input)
    expect(output).toEqual('')
  })
})
