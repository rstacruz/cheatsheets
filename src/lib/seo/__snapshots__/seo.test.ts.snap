// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getSEOPropsForPage() > title only 1`] = `
{
  "links": {
    "canonical": "https://devhints.io/react",
  },
  "meta": {
    "app:pageurl": "https://devhints.io/react",
    "description": "The one-page guide to React: usage, examples, links, snippets, and more.",
  },
  "metaProperties": {
    "article:tag": [],
    "og:description": "The one-page guide to React: usage, examples, links, snippets, and more.",
    "og:image": "https://assets.devhints.io/previews/react.jpg",
    "og:image:height": "471",
    "og:image:width": "900",
    "og:site_name": "Devhints.io cheatsheets",
    "og:title": "React cheatsheet",
    "og:type": "article",
    "og:url": "https://devhints.io/react",
    "twitter:description": "The one-page guide to React: usage, examples, links, snippets, and more.",
    "twitter:image": "https://assets.devhints.io/previews/react.jpg",
    "twitter:title": "React cheatsheet",
  },
  "title": "React cheatsheet",
}
`;

exports[`getSEOPropsForPage() > with Markdown description 1`] = `
{
  "links": {
    "canonical": "https://devhints.io/react",
  },
  "meta": {
    "app:pageurl": "https://devhints.io/react",
    "description": "A React cheatsheet (Markdown text) · One-page guide to React",
  },
  "metaProperties": {
    "article:tag": [],
    "og:description": "A React cheatsheet (Markdown text) · One-page guide to React",
    "og:image": "https://assets.devhints.io/previews/react.jpg",
    "og:image:height": "471",
    "og:image:width": "900",
    "og:site_name": "Devhints.io cheatsheets",
    "og:title": "React cheatsheet",
    "og:type": "article",
    "og:url": "https://devhints.io/react",
    "twitter:description": "A React cheatsheet (Markdown text) · One-page guide to React",
    "twitter:image": "https://assets.devhints.io/previews/react.jpg",
    "twitter:title": "React cheatsheet",
  },
  "title": "React cheatsheet",
}
`;

exports[`getSEOPropsForPage() > with description 1`] = `
{
  "links": {
    "canonical": "https://devhints.io/react",
  },
  "meta": {
    "app:pageurl": "https://devhints.io/react",
    "description": "A React cheatsheet · One-page guide to React",
  },
  "metaProperties": {
    "article:tag": [],
    "og:description": "A React cheatsheet · One-page guide to React",
    "og:image": "https://assets.devhints.io/previews/react.jpg",
    "og:image:height": "471",
    "og:image:width": "900",
    "og:site_name": "Devhints.io cheatsheets",
    "og:title": "React cheatsheet",
    "og:type": "article",
    "og:url": "https://devhints.io/react",
    "twitter:description": "A React cheatsheet · One-page guide to React",
    "twitter:image": "https://assets.devhints.io/previews/react.jpg",
    "twitter:title": "React cheatsheet",
  },
  "title": "React cheatsheet",
}
`;

exports[`getSEOPropsForPage() > with description, keywords 1`] = `
{
  "links": {
    "canonical": "https://devhints.io/react",
  },
  "meta": {
    "app:pageurl": "https://devhints.io/react",
    "description": "A React cheatsheet · One-page guide to React",
  },
  "metaProperties": {
    "article:tag": [],
    "og:description": "A React cheatsheet · One-page guide to React",
    "og:image": "https://assets.devhints.io/previews/react.jpg",
    "og:image:height": "471",
    "og:image:width": "900",
    "og:site_name": "Devhints.io cheatsheets",
    "og:title": "React cheatsheet",
    "og:type": "article",
    "og:url": "https://devhints.io/react",
    "twitter:description": "A React cheatsheet · One-page guide to React",
    "twitter:image": "https://assets.devhints.io/previews/react.jpg",
    "twitter:title": "React cheatsheet",
  },
  "title": "React cheatsheet",
}
`;

exports[`getSEOPropsForPage() > with keywords 1`] = `
{
  "links": {
    "canonical": "https://devhints.io/react",
  },
  "meta": {
    "app:pageurl": "https://devhints.io/react",
    "description": "hooks · components · props · One-page guide to React",
  },
  "metaProperties": {
    "article:tag": [],
    "og:description": "hooks · components · props · One-page guide to React",
    "og:image": "https://assets.devhints.io/previews/react.jpg",
    "og:image:height": "471",
    "og:image:width": "900",
    "og:site_name": "Devhints.io cheatsheets",
    "og:title": "React cheatsheet",
    "og:type": "article",
    "og:url": "https://devhints.io/react",
    "twitter:description": "hooks · components · props · One-page guide to React",
    "twitter:image": "https://assets.devhints.io/previews/react.jpg",
    "twitter:title": "React cheatsheet",
  },
  "title": "React cheatsheet",
}
`;
